src/weights.jl

```diff
@@ -25,13 +25,9 @@ isempty(wv::AbstractWeights) = isempty(wv.values)
 Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
 Base.size(wv::AbstractWeights) = size(wv.values)
 
-function Base.setindex!(wv::AbstractWeights{S, T}, v::T, i::Int) where {S, T}
-    s = sum(wv)
-    s -= wv[i]
-    s += v
-
+@propagate_inbounds function Base.setindex!(wv::AbstractWeights, v::Real, i::Int)
+    wv.sum += v - wv[i]
     wv.values[i] = v
```

Comment:
Actually I realize the assignment should be done first, so that if it fails the sum isn't modified. But you still need to return `v` explicitly (cf. comment below).